File: ProfilerHost/MetadataRegistry.cpp
#include "stdafx.h"
#include "MetadataRegistry.h"

MetadataRegistry::MetadataRegistry()
    : tokenCounter(0)
{
}

MetadataRegistry::~MetadataRegistry()
{
}

MethodMetadata MetadataRegistry::RegisterMethod(
    std::wstring assemblyName,
    std::wstring typeName,
    std::wstring signature)
{
    MethodMetadata metadata;
    metadata.token = InterlockedIncrement64((volatile LONG64*)(&this->tokenCounter));
    metadata.assemblyName = assemblyName;
    metadata.typeName = typeName;
    metadata.signature = signature;

    this->methodMetadatas.push_back(metadata);
    if (this->methodRegistrationCallback)
    {
        this->methodRegistrationCallback(metadata);
    }

    return metadata;
}

void MetadataRegistry::SetMethodRegistrationCallback(std::function<void(const MethodMetadata&)> callback)
{
    this->methodRegistrationCallback = callback;
}

File: ProfilerCore/MetadataRegistry.cs
namespace ProfilerCore
{
    using System.Collections.Concurrent;
    using System.Collections.Generic;

    internal sealed class MethodMetadata
    {
        public ulong Token { get; set; }
        public string AssemblyName { get; set; }
        public string TypeName { get; set; }
        public string Signature { get; set; }
    }

    internal sealed class MetadataRegistry
    {
        private readonly ConcurrentDictionary<ulong, MethodMetadata> methodMetadatas;

        public MetadataRegistry()
        {
            this.methodMetadatas = new ConcurrentDictionary<ulong, MethodMetadata>();
        }

        public void RegisterMethod(MethodMetadata method)
        {
            this.methodMetadatas[method.Token] = method;
        }

        public MethodMetadata ResolveMethodToken(ulong token)
        {
            MethodMetadata metadata;
            return !this.methodMetadatas.TryGetValue(token, out metadata) ? null : metadata;
        }
    }
}
